content/7-webapp-features/1-fetch-data/angular/user.service.ts

```diff
@@ -1,25 +1,57 @@
 import { HttpClient } from '@angular/common/http';
 import { Injectable } from '@angular/core';
-import { catchError, map, Subject } from 'rxjs';
-import { User } from './user';
+import { BehaviorSubject, catchError, take } from 'rxjs';
 
-@Injectable({
-  providedIn: 'root'
-})
+export interface UsersState {
+	users: User[];
+	error: string | null;
+	loading: boolean;
+}
+
+export const initialState: UsersState = {
+	users: [],
+	error: null,
+	loading: false,
+};
 
+@Injectable({ providedIn: 'root' })
 export class UserService {
-  error$: Subject<any> = new Subject();
-  
-  private users$ = this.HttpClient.get<{results: User[]; info: any}>('https://randomuser.me/api/?results=3')
-  .pipe(
-    map(x => x.results),
-    catchError(e => {this.error$.next(e); return []})
-  );
-  
-  constructor(private HttpClient: HttpClient) {}
-
-  getUsers() {
-    return this.users$;
-  }
+	private state = new BehaviorSubject<UsersState>(initialState);
+	state$ = this.state.asObservable();
+
+	constructor(private http: HttpClient) {}
+
+	loadUsers() {
+		this.state.next({ ...initialState, loading: true });
+
+		this.http
+			.get<UserRes>('https://randomuser.me/api/?results=3')
+			.pipe(
+				take(1),
+				tap(({ results }) => this.state.next({ ...initialState, users: results })),
+				catchError((error) => {
+					this.state.next({ ...initialState, error });
+					return error;
+				})
+			)
+			.subscribe();
+	}
+}
+
+export interface UserRes {
```

Comment:
I'm not quite sure a regular JavaScript or Angular developer can understand this easily.
And is `rxjs` always used in Angular apps ?

Reply:
A regular Angular developer can understand that easily as it's just a  **Service with a Subject** and it's common in Angular world. 

And yes, rxjs is always used in Angular apps. Even the http.get() returns an rxjs Observable.